exercise_2/openweather_api_client.py

```diff
@@ -0,0 +1,83 @@
+import requests
+import collections
+
+
+WeatherForecast = collections.namedtuple(
+    "WeatherForecast", ("weather_forecast_type", "city", "country", "current_weather", "temperature", "pressure")
+    )
+
+
+class BasicWeatherApiClient:
+    """
+    """
+    data_class = None
+
+    def get_weather(self, type, value, api_key):
+        """Get a weather forecast from openweathermap.org.
+
+        Parameters:
+        type (string): city identification type. Available types:
+            q - city name
+            zip - zip code
+            id - city id
+
+        value (string): value determined by type. Available values (examples):
+            q = London
+            zip = 94040
+            id = 2172797
+
+        api_key (string): API key generated by openweathermap.org
+
+        """
+        pass
+
+
+class CurrentWeatherApiClient(BasicWeatherApiClient):
+    data_class = WeatherForecast
+
+    def get_weather(self, type, value, api_key):
+        url = "http://api.openweathermap.org/data/2.5/weather?{}={}&APPID={}".format(type, value, api_key)
+
+        weather_information = requests.get(url).json()
+        data = {
+            "weather_forecast_type": "Current weather",
+            "city": weather_information.get("name"),
+            "country": weather_information["sys"].get("country"),
+            "current_weather": weather_information["weather"][0].get("main"),
+            "temperature": weather_information["main"].get("temp"),
+            "pressure": weather_information["main"].get("pressure"),
+        }
+        return self.data_class(**data)
+
+
+class FiveDayForecastApiClient(BasicWeatherApiClient):
+    data_class = WeatherForecast
+
+    def get_weather(self, type, value, api_key):
+        url = "http://api.openweathermap.org/data/2.5/forecast?{}={}&APPID={}".format(type, value, api_key)
+
+        city_information = requests.get(url).json()["city"]
+        weather_information = requests.get(url).json()["list"][0]
+        data = {
+            "weather_forecast_type": "5 days forecast",
+            "city": city_information.get("name"),
+            "country": city_information.get("country"),
+            "current_weather": weather_information["weather"][0].get("main"),
+            "temperature": weather_information["main"].get("temp"),
+            "pressure": weather_information["main"].get("pressure"),
+        }
+        return self.data_class(**data)
+
+
+class WeatherForecastManager:
```

Comment:
Jest z tym wszystkim jeden problem. Mianowicie zrobiłaś manager'a, który przy odpaleniu run, zwraca dwie pogody. Nie chodziło tak naprawdę o to. Dobrze nadziałaś się na nadużywanie dziedziczenia, gdzie powinnaś użyć kompozycji. :D APIClient'y mają trochę inne zasady działania i design. Mianowicie:
`1. Inicjalizujesz clienta, podając jakiś klucz albo inne dane, które służą do autentykacji żeby nie przekazywac tego cały czas.`
`2. Wykonujesz jakąś metodę na API, która cuś robi i zwraca dane przy czym te dane sprowadza w jakimś cywilizowanym formacie.`
`3. Zwrotkę wykorzystujesz w swoim kodzie do np. powiadomienia klienta, że udało się wrzucić dane czy coś innego`

APIClient'y są tak gotową biblioteką/modułem, który obsługuje podstawowe operacje na jakimś Web'owym (zwykle REST'owym API). REST'owe API wyróżniają się w swoim design'ie tym, że końcówki (endpoint'y) odpowiadają za konkretne resource'y - tzn.:
`api/toys` - wiemy, że taka końcówka ma coś wspólnego z zabawkami, wykonanie na niej GET prawdpodobnie zwróci  dane o zabawkach.
`api/toys/sale` - taka końcówka zwróciłaby pewnie dane o zabawkach z przeceny.
Używając tego przykładu jako zrozumienia design'u i struktury byłoby.
`client = NonameAPIClient()`
`client.toys.get() zwróciloby info z api/toys/`
`client.toys.get_toys_from_sale() zwrociloby z api/toys/sale` Rozumisz ?

 Preferowanym design'em (wspomogę Cię) z strukturą logiczną odpowiadającą też temu REST'owemu design'owi byłoby coś takiego:
`openweather_client = OpenWeatherAPIClient(api_key=config.OPENWEATHER_API_KEY)`
`current = openweather_client.current_weather.get(q=London)   # przekazywane do zapytania jako kwargs np`
`forecast = openweather_client.five_days_forecast.get(q=London)`
W taki sposób oddzielasz implementacje końcówek od siebie (przez kompozycję `current_weather` `five_days_forecast`). Końcówki mogą mieć jakąś cześć wspólną (klasa bazowa, jakiś interfejs)
W obrębie jednej końcówki tzn. current_weather może być wiele metod tzn. current_weather.get current_weather.cos_innego ale musi to oczywiście odpowiadać temu na co pozwala openweather api. Nie mówię, że ten dizajn jest najlepszy ale jest całkiem logiczny i odpowiadający temu jak pisane są REST API. Co myślisz ? :D